src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,180 @@
+                index < HEAD_INDEX ? index : index + ACCURACY_OF_ONE);
+        Node targetNode = findNodeByIndex(index);
+        return (T) targetNode.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIfIndexOutOfBoundsException(MESSAGE_OF_SET_VALUE_EXCEPTION,
+                index < HEAD_INDEX ? index : index + ACCURACY_OF_ONE);
+        Node targetNode = findNodeByIndex(index);
+        T targetNodeValue = (T) targetNode.value;
+        targetNode.value = value;
+        return targetNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIfIndexOutOfBoundsException(MESSAGE_OF_REMOVE_VALUE_EXCEPTION,
+                index < HEAD_INDEX ? index : index + ACCURACY_OF_ONE);
+        Node targetNode = findNodeByIndex(index);
+        if (index == HEAD_INDEX) {
+            head = targetNode.next;
+        }
+        if (index == usedSpace) {
+            tail = targetNode.prev;
+        }
+        unlink(targetNode);
+        usedSpace--;
+        return (T) targetNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        for (int i = 0; i < usedSpace; i++) {
+            if (currentNode.value == object
+                    || currentNode.value != null && currentNode.value.equals(object)) {
+                if (i == HEAD_INDEX) {
+                    head = currentNode.next;
+                }
+                unlink(currentNode);
+                usedSpace--;
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return usedSpace;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return usedSpace == EMPTY;
+    }
+
+    private void unlink(Node node) {
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        }
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        }
+    }
+
+    private Node findNodeByIndex(int index) {
+        int halfOfUsedSpace = (int) Math.ceil((double) usedSpace / 2);
+        if (index <= halfOfUsedSpace) {
+            Node currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+            return currentNode;
+        } else {
+            Node currentNode = tail;
+            for (int i = usedSpace - ACCURACY_OF_ONE; i > index; i--) {
+                currentNode = currentNode.prev;
+            }
+            return currentNode;
+        }
     }
+
+    private void checkIfIndexOutOfBoundsException(String message, int index) {
+        if (index > usedSpace || index < HEAD_INDEX) {
+            throw new IndexOutOfBoundsException(message + (usedSpace - ACCURACY_OF_ONE));
+        }
+    }
+
+    private class Node<V> {
```

Comment:
Why you don't use `private class Node<T>`?

Reply:
because it stores the V value 😅